cpp/src/arrow/util/async_generator.h

```diff
@@ -1645,6 +1667,47 @@ AsyncGenerator<T> MakeCancellable(AsyncGenerator<T> source, StopToken stop_token
   return CancellableGenerator<T>{std::move(source), std::move(stop_token)};
 }
 
+template <typename T>
+struct PauseableGenerator {
+ public:
+  PauseableGenerator(AsyncGenerator<T> source, std::shared_ptr<util::AsyncToggle> toggle)
+      : state_(std::make_shared<PauseableGeneratorState>(std::move(source),
+                                                         std::move(toggle))) {}
+
+  Future<T> operator()() { return (*state_)(); }
+
+ private:
+  struct PauseableGeneratorState
+      : public std::enable_shared_from_this<PauseableGeneratorState> {
+    PauseableGeneratorState(AsyncGenerator<T> source,
+                            std::shared_ptr<util::AsyncToggle> toggle)
+        : source_(std::move(source)), toggle_(std::move(toggle)) {}
+
+    Future<T> operator()() {
+      std::shared_ptr<PauseableGeneratorState> self = this->shared_from_this();
+      return toggle_->WhenOpen().Then([self] {
```

Comment:
Below there's a comment about how later calls to Open may finish before earlier calls - doesn't this mean that this generator can get inadvertently reordered?

(Also, it relies on Future running callbacks in order, right?)

Reply:
Yes, the generator can reorder things.  If we want order we'd have to rely on the future running callbacks in order (although that should be safe to rely on).  If we don't care about order we don't have to rely on that.

At the time I didn't think reordering would be a bad thing since exec plan is likely to reorder stuff anyways but I don't really like it.  I might take a look to see if I can easily improve it.

Reply:
I guess my main concern is if the end token gets reordered - or is that not a problem?

Reply:
I cleaned it up a little bit which should eliminate most reordering but I could not escape the case:

Thread A calls Open
Thread D calls WhenOpen
Thread B calls Close
Thread C calls Open
Thread D calls WhenOpen

It is possible for callbacks on the second call to WhenOpen to finish before callbacks on the first call.

So what I did was go ahead and change the comments for MakePauseable to state that it is not async reentrant.  This should prevent any possible reordering because the second call to WhenOpen will not happen until the future returned from the first call to WhenOpen has completed.  The only spot we are using MakePauseable does not need async reentrancy so this should be ok.

Reply:
Sounds good.